misk/src/main/kotlin/misk/tasks/RepeatedTaskQueue.kt

```diff
@@ -0,0 +1,126 @@
+  private val dispatchExecutor: Executor?, // visible internally for testing only
+  private val pendingTasks: BlockingQueue<DelayedTask> // visible internally for testing only
+) : AbstractExecutionThreadService() {
+  constructor(name: String, clock: Clock, taskExecutor: ExecutorService) :
+      this(name, clock, taskExecutor, null, DelayQueue<DelayedTask>())
+
+  private val running = AtomicBoolean(false)
+
+  override fun startUp() {
+    log.info { "starting repeated task queue $name" }
+    running.set(true)
+  }
+
+  override fun shutDown() {
+    log.info { "stopping repeated task queue $name" }
+
+    // Remove all currently scheduled tasks, and schedule an empty task to kick the background thread
+    pendingTasks.clear()
+    pendingTasks.add(DelayedTask(clock, clock.instant()) {
+      Result(Status.OK_NO_RESCHEDULE, Duration.ofMillis(0))
+    })
+  }
+
+  /**
+   * runs the main event loop, pulling the next task from the queue and handing it off to the
+   * executor for dispatching
+   */
+  override fun run() {
+    while (running.get()) {
+      // Fetch the next task, bailing out if we've shutdown
+      val task = pendingTasks.take().task
+      if (!running.get()) {
+        return
+      }
+
+      // Hand the task off to the executor for parallel execution and repeat so long as the
+      // task requests rescheduling
+      taskExecutor.submit {
+        val result = task()
+        if (!result.status.isFinalExecution) schedule(result.nextDelay, task)
+      }
+    }
+  }
+
+  /**
+   * Schedules a task to run repeatedly after an initial delay. The task itself determines the
+   * next execution time
+   */
+  fun schedule(delay: Duration, task: () -> Result) {
+    pendingTasks.add(DelayedTask(clock, clock.instant().plus(delay), task))
+  }
+
+  /**
+   * Schedules a task to run repeatedly at a fixed delay, with back-off for errors and lack
+   * of available work
+   */
+  fun scheduleWithBackoff(
+    timeBetweenRuns: Duration,
+    initialDelay: Duration = timeBetweenRuns,
+    noWorkBackoff: Backoff = ExponentialBackoff(timeBetweenRuns, defaultMaxDelay, defaultJitter),
+    failureBackoff: Backoff = ExponentialBackoff(timeBetweenRuns, defaultMaxDelay, defaultJitter),
+    task: () -> Status
+  ) {
+    val wrappedTask: () -> Result = {
+      try {
+        val status = task()
+        when (status) {
+          Status.OK -> {
+            noWorkBackoff.reset()
+            failureBackoff.reset()
+            Result(status, timeBetweenRuns)
+          }
+          Status.NO_WORK -> {
+            failureBackoff.reset()
+            Result(status, noWorkBackoff.nextRetry())
+          }
+          Status.FAILED -> {
+            noWorkBackoff.reset()
+            Result(status, failureBackoff.nextRetry())
+          }
+          Status.OK_NO_RESCHEDULE, Status.FAILED_NO_RESCHEDULE ->
+            Result(status, Duration.ofMillis(0))
+        }
+      } catch (th: Throwable) {
+        noWorkBackoff.reset()
```

Comment:
will this exception be logged in any way?

Reply:
It's up to the task to log - we lack sufficient context to do so here

Reply:
We'll backup log here; I'm generally opposed to low level libraries logging because they don't have context and tend to make things extra chatty, but in this place it's probably important enough to do

Reply:
agreed. it seems at this point there is a bug in the task handler since it did not properly return `FAILED`